### PR TITLE
Load JavaScript modules from Node packages

### PR DESCRIPTION
**Pull request by billti**
I'm not looking to push this now, but opening for discussion on both a) the approach in the implementation, and b) some design questions it raises.

This change would allow for the loading of JavaScript modules from a search of top level Node packages. This could be useful in TypeScript, but is especially useful in editing JavaScript.

To avoid transitively loading 100s of JavaScript files from an NPM installed dependency graph, this sets a limit on how deep in the dependency chain to crawl (which only applies to JavaScript, not TypeScript files). For example, if `chalk` depends on `has-ansi` which depends on `ansi-regex`, then setting `maxNodeModuleJsDepth` to 0 (the default currently) wouldn't load JavaScript modules for any of them. Setting to 1 would load only the JavaScript modules for `chalk`, to 2 would load `has-ansi` also, etc.. If a module exposes imports from a dependency, then a greater search depth flows up richer type information (e.g. packages such as `gulp` commonly reexport members of their imported dependencies, such as `vinyl-fs`).

In order to avoid recompiling/overwriting JavaScript files installed under `node_modules`, this change also suppresses emitting any code for files found with a search of installed Node packages. This is a change in behavior (as can be seen in some baseline changes below), but what is desirable here is up for debate. It seems bad if you end up renaming a property in a dependency you installed from NPM (which will be reverted on next `npm install`, likely resulting in a mismatch of property names), yet some folks also `npm link` across multiple projects under development in order to treat the whole as a "solution".

@vladima The approach below no longer modifies the `ResolvedModule` type (it uses the existing `isExternalLibraryImport` flag), but does add a member to `SourceFile`. Let me know if this requires managed side changes also.

### CODE REVIEW

**Comment by Bigous** [comment]
This is awesome, I think it could be added to 1.9 or 1.8 ... very, very useful.
I think depth 2 would be enough for the majority. So it could be the default.
Cache the findings of Salsa as a pseudo .d.ts (or a real .d.ts  in typings/generated) could be great too.
Another approach could be a simple line command option to generate all .d.ts from node_modules that does not have .d.ts or .ts itself. It could help people to write the .d.ts from packages without having to start from 0.

**Comment by massimocode** [comment]
I've not looked at the code but this looks like a very sensible/useful idea

File: src/compiler/program.ts [hunk 1]
```diff
@@ -79,9 +79,7 @@ namespace ts {
             return false;
         }
 
-        const i = moduleName.lastIndexOf("./", 1);
-        const startsWithDotSlashOrDotDotSlash = i === 0 || (i === 1 && moduleName.charCodeAt(0) === CharacterCodes.dot);
-        return !startsWithDotSlashOrDotDotSlash;
+        return !startsWithDotSlashOrDotDotSlash.test(moduleName);
```
Comment (mhegazy) [inline src/compiler/program.ts]: The idea here is that looking up the fist 2 chars is cheeper than running a regexp. is there a reason why we need to change it?

File: src/compiler/program.ts [hunk 2]
```diff
@@ -1468,11 +1489,13 @@ namespace ts {
         }
 
         function processImportedModules(file: SourceFile, basePath: string) {
+            const maxJsNodeModuleSearchDistance = options.maxNodeModuleJsDepth || 0;
```
Comment (mhegazy) [inline src/compiler/program.ts]: so is the default is meant to be `0` or are we going to give it a higher value and users can opt in to increase it if they do not see that fit?

File: src/compiler/utilities.ts [hunk 3]
```diff
@@ -2035,7 +2035,8 @@ namespace ts {
         else {
             const sourceFiles = targetSourceFile === undefined ? host.getSourceFiles() : [targetSourceFile];
             for (const sourceFile of sourceFiles) {
-                if (!isDeclarationFile(sourceFile)) {
+                // Don't emit if source file is a declaration file, or was found by a search under 'node_modules'
+                if (!isDeclarationFile(sourceFile) && !sourceFile.nodeModuleSearchDistance) {
```
Comment (mhegazy) [inline src/compiler/utilities.ts]: we should add a test with --noResolve for this.

**Comment by mhegazy** [comment]
a few minor comments. other than that looks good to me.

**Comment by raybooysen** [comment]
@mhegazy Is this still on for a 2.0 release?  Is it too late for a 1.9?

**Comment by goldenc** [comment]
I could really use this getting into the 1.9 release.. pretty please!

**Comment by dunkymole** [comment]
a 1.9 release would help us tremendously also

**Comment by jonclare** [comment]
+1 for getting this in the 1.9 release - would be a great help for us.

**Comment by bnwan** [comment]
It would be great if this made it to 1.9

**Comment by sarbbottam** [comment]
@billti could you resolve the conflicts?

When could this PR be merged? I look forward to see https://github.com/Microsoft/TypeScript/issues/6670 resolved 

cc/ egamma 

Thanks!

**Comment by mhegazy** [comment]
@billti is this ready to go? can you update?

**Comment by billti** [comment]
I'll try and get to it later tonight or tomorrow... just got a couple of other tasks ahead of it.

**Comment by billti** [comment]
Well, that was quite a painful merge :tired_face: , but this should be up to date with `master` now and passing the tests.

Seeing as the module resolution changed so much since 1.x, I'll spend some more time on exploratory and automated testing, but it'd be great if you ( @vladima ) could give it the once over for any obvious problems also.

**Comment by mhegazy** [comment]
One comment, we cannot set `nodeModuleSearchDistance` on the source file, it has to be set on the program. the reason is a SourceFile object can be shared between multiple projects, some as import, some as `/// <reference>`, how the file was resolved is function of the program.

**Comment by billti** [comment]
They say necessity is the mother of invention... by rethinking how to handle this with the new requirement that a SourceFile might be reused across programs, I think this is actually much simpler now. 

@mhegazy, can you take a look and let me know if you spot any flaws with this approach?

I don't handle the suppression of emit if the file is located under node_modules, but that is a separate issue that is present already. I can add that and more tests if desired before checking in.

**Comment by billti** [comment]
Actually, trying to solve the emit problem I realized where this falls down in the general case.

If module A imports B imports C, and is doesn't add C when processing B because it's too many hops, if I then later directly import B, is will already have processed the file and won't reprocess imports and see that C is now not too far and should be added. :cry: Back to the drawing board...

**Comment by mhegazy** [comment]
👍

**Comment by mhegazy** [comment]
Can you also add a test for `--maxNodeModuleJsDepth`

**Comment by billti** [comment]
@mhegazy This should be pretty comprehensive now. It tracks if a JavaScript file was loaded from searching node_modules so it knows not to compile it (a minor tweak to this logic could solve #6964 if desired). It also tracks if a module has some imports skipped, so that if the same module is later processed higher up the chain it will re-processes the imports.

I added a few project tests to test the eliding once the max depth is reached, increasing the depth via the config option, and the reprocessing on a higher up import. I used Project tests as with baselines its hard to add files that exist but aren't root files (as the files under node_modules need to be). The project tests also contain useful info about what files were input, resolved, and output.

Let me know if you have any other suggestions for improvements.

File: tests/baselines/reference/library-reference-5.errors.txt [hunk 4]
```diff
@@ -18,7 +18,7 @@
 ==== /node_modules/bar/index.d.ts (1 errors) ====
     /// <reference types="alpha" />
     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
-!!! message TS4090: Conflicting library definitions for 'alpha' found at 'index.d.ts' and 'index.d.ts'. Copy the correct file to the 'typings' folder to resolve this conflict.
+!!! message TS4090: Conflicting library definitions for 'alpha' found at '/node_modules/bar/node_modules/alpha/index.d.ts' and '/node_modules/foo/node_modules/alpha/index.d.ts'. Copy the correct file to the 'typings' folder to resolve this conflict.
```
Comment (billti) [inline tests/baselines/reference/library-reference-5.errors.txt]: Fixing the regular expression makes messages such as this useful.

File: src/compiler/program.ts [hunk 5]
```diff
@@ -1334,7 +1361,10 @@ namespace ts {
                 getNewLine: () => host.getNewLine(),
                 getSourceFile: program.getSourceFile,
                 getSourceFileByPath: program.getSourceFileByPath,
-                getSourceFiles: program.getSourceFiles,
+                getSourceFiles: () => filter(program.getSourceFiles(),
+                    // Remove JavaScript files found by searching node_modules from the source files to emit
+                    sourceFile => !lookUp(jsFilesFoundSearchingNodeModules, sourceFile.path)
```
Comment (mhegazy) [inline src/compiler/program.ts]: do not think we should do this. the file is part of the program. getSourceFle, and getSourceFile, should give an accurate representation of the files in the program. you could get an error in this file, or get a definition in this file, and using the file name you should be able to get to the SourceFile object using `getSourceFile` and you should also be able to find it if you enumerate the `getSourceFiles`.

I would say create a new method, `Program.jsFilesFoundSearchingNodeModules(path)` that the emitter checks and skips emitting if true.

**Comment by mhegazy** [comment]
👍

**Comment by billti** [comment]
How about this change? 

Note that I'm using a Map for the collection of file names as this seems quicker on large projects with lots of modules from node_modules, than using a array of file paths and scanning this every time to check for existence (and splicing when I need to remove an entry). I could use a simple string[] if that seems cleaner, but this works well currently.

**Comment by mhegazy** [comment]
I would have preferred a function that takes a file path and returns a boolean. but this is fine too.

**Comment by mhegazy** [comment]
can you mark this as `/* @internal */` since we are the only users for it.

**Comment by billti** [comment]
Like so?

**Comment by mhegazy** [comment]
thanks!

**Comment by mhegazy** [comment]
no objections. thanks!

**Comment by sarbbottam** [comment]
👏 @billti @mhegazy

**Pull request merged**